src/build_queue.rs

```diff
@@ -343,121 +372,158 @@ impl AsyncBuildQueue {
-                let priority = get_crate_priority(&mut conn, &release.name).await?;
-
-                match self
-                    .add_crate(
-                        &release.name,
-                        &release
-                            .version
-                            .parse()
-                            .context("couldn't parse release version as semver")?,
-                        priority,
-                        index.repository_url(),
-                    )
-                    .await
-                    .with_context(|| {
-                        format!(
-                            "failed adding {}-{} into build queue",
-                            release.name, release.version
-                        )
-                    }) {
-                    Ok(()) => {
-                        debug!(
-                            "{}-{} added into build queue",
-                            release.name, release.version
-                        );
-                        self.queue_metrics.queued_builds.add(1, &[]);
+    async fn process_changes(
+        &self,
+        conn: &mut AsyncPoolClient,
+        changes: &Vec<Change>,
+        registry: Option<&str>,
+    ) -> usize {
+        let mut crates_added = 0;
+
+        for change in changes {
+            match self.process_change(conn, change, registry).await {
+                Ok(added) => {
+                    if added {
                         crates_added += 1;
                     }
-                    Err(err) => report_error(&err),
                 }
+                Err(err) => report_error(&err),
             }
+        }
+        crates_added
+    }
 
-            let yanked = change.yanked();
-            let unyanked = change.unyanked();
-            if let Some(release) = yanked.or(unyanked) {
-                // FIXME: delay yanks of crates that have not yet finished building
-                // https://github.com/rust-lang/docs.rs/issues/1934
-                if let Ok(release_version) = Version::parse(&release.version)
-                    && let Err(err) = self
-                        .set_yanked_inner(
-                            &mut conn,
-                            release.name.as_str(),
-                            &release_version,
-                            yanked.is_some(),
-                        )
-                        .await
-                {
-                    report_error(&err);
-                }
-
-                self.queue_crate_invalidation(&release.name).await;
+    /// Process a crate change, returning whether the change was a crate addition or not.
+    async fn process_change(
+        &self,
+        conn: &mut AsyncPoolClient,
+        change: &Change,
+        registry: Option<&str>,
+    ) -> Result<bool> {
+        match change {
+            Change::Added(release) => self.process_version_added(conn, release, registry).await?,
+            Change::AddedAndYanked(release) => {
```

Comment:
there is still a difference in error handling compared to before, for this variant. 

old behaviour: we would still process the yank change, even when the "added" event failed. 
New behaviour wouldn't do it. 

But looking at past sentry errors I think we're good.